server/routerlicious/packages/lambdas/src/deli/lambdaFactory.ts

```diff
@@ -93,28 +113,46 @@ export class DeliLambdaFactory
 		};
 
 		let gitManager: IGitManager;
-		let document: IDocument;
+		let document: IDocument = {} as IDocument;
 
 		try {
+			const documentCacheKey = invalidDocumentCacheKey(tenantId, documentId);
+			const sessionCacheKey = invalidSessionCacheKey(tenantId, documentId);
+			const invalidDocumentCacheEntryExists = this.invalidDocumentOrSessionCacheEnabled ? this.invalidDocumentOrSessionCache.has(documentCacheKey) : false;
+			const invalidSessionCacheEntryExists = this.invalidDocumentOrSessionCacheEnabled ? this.invalidDocumentOrSessionCache.has(sessionCacheKey) : false;
+
+			const lumberjackProperties = {
+				...getLumberBaseProperties(documentId, tenantId),
+				invalidDocumentOrSessionCacheEnabled: this.invalidDocumentOrSessionCacheEnabled,
+				invalidDocumentCacheEntryExists,
+				invalidSessionCacheEntryExists,
+			};
+			Lumberjack.info("Cached invalid document or session.", lumberjackProperties);
+
 			// Lookup the last sequence number stored
 			// TODO - is this storage specific to the orderer in place? Or can I generalize the output context?
-			document = await this.documentRepository.readOne({ documentId, tenantId });
+			if (!invalidDocumentCacheEntryExists && !invalidSessionCacheEntryExists) {
```

Comment:
Would the cache reset on pod restarts? If yes, does it still help in other situations we might have invalid sessions/documents?

Reply:
Cache will reset on pod restarts. However, telemetry shows that for a lot of documents this happens on a given pod over 100K times during a span of several hours. So, this still should be helpful.

With Redis it will be more efficient, however we will add extra load on Redis. Starting with in-memory cache is a trade-off. We can see how effective is this and switch to Redis as necessary.

Reply:
Got it. Do those documents come back up within the same pod in that few hours timespan, or were there restarts in that period?

The reason I ask is: I was under the impression that even within a few hours period, the pods were restarting quite a bit. So I was expecting this to happen: one original restart, which would reset the cache, causes a lot of invalid documents to be brought back to life. That leads to a lot of trouble, eventually causing the pod to crash and restart again. That resets the cache again, and then, when the pod comes back up, the same documents would show up and not be ignored because the cache was reset.

I also think having Redis here might require some more work/research to analyze impact. I was approaching this from a different angle: what risks it can bring if the behavior I described above is confirmed and the feature does not help as expected.

Unfortunately I think there might be no straightforward way to test this... :(

Reply:
I think that when the pod restarts and we have lots of invalid docs, the reason it is restarting again is due to the very large spike in the outgoing mongo requests which is causing both outgoing mongo requests and kafka checkpoints to fail. If we do not have the mongo spike, there is a chance that we might not even go into a restart loop (unless it is a oom but in that case, redis is not really going to help)

Reply:
My question is: do we see the same invalid document ID showing up more than once before a pod crash/restart happens? If we see the same document ID more than once and the cache is not flushed, then I totally agree this would help. My concern was because I thought that was not the case; instead, I assumed that the spike of invalid documentIDs that happen upon each restart only includes unique documentIDs. In case that's true and they are all unique, caching them will just add more memory pressure and we might not get the benefit of avoiding the outgoing network calls because there is never going to be a cache hit

Reply:
Pradeep mentioned to me in a different chat:
> even with this change the number of mongo calls will reduce by the avg number of ops getting replayed per doc which should be a substantial reduction..it would be ideal if we use redis but this is still definitely worth trying since it is comparatively simpler to implement

So this means that my assumption above was wrong, and that each documentID shows up multiple times during these spikes. So my comment does not make sense. I agree this will be helpful. 😄